### PR TITLE
Intrinsic operator support in template instantiation

### PR DESCRIPTION
**Pull request by ansharlubis**
Started working on supporting intrinsic operator in template instantiation (#1199). Operator works only as alias for now.

### CODE REVIEW

File: integration_tests/template_add_02.f90 [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    real function func_arg_real(x, y) result(z)
+        real, intent(in) :: x, y
+        z = x + y
+    end function
+
+    integer function func_arg_int(x, y) result(z)
+        integer, intent(in) :: x, y
+        z = x + y
+    end function
```
Comment (certik) [inline integration_tests/template_add_02.f90:40]: These can now be removed I think.

Comment (ansharlubis) [reply integration_tests/template_add_02.f90:40]: Not yet I'm afraid, the operator currently only works as alias for functions defined in the interface.

File: tests/tests.toml [hunk 2]
```diff
@@ -13,18 +13,23 @@ ast = true
...
 asr = true
 
 [[test]]
 filename = "../integration_tests/template_nested.f90"
-ast = true
 asr = true
 
 [[test]]
 filename = "../integration_tests/template_travel.f90"
-ast = true
```
Comment (certik) [inline tests/tests.toml:27]: I would leave these two AST tests there, just to check that the AST looks good.

**Comment by certik** [comment]
@ansharlubis is this ready for review? Overall it looks good to me.

**Comment by ansharlubis** [comment]
@certik I'd like to review it myself first, I'll let you know later.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 3]
```diff
@@ -510,37 +510,177 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                        false, false, false, nullptr, 0, nullptr, 0, false, false, true);
+                    
+                    ASR::symbol_t *op_sym = ASR::down_cast<ASR::symbol_t>(op_function);
+                    parent_scope->add_symbol(func_name, op_sym);
+                    current_scope = parent_scope;
+                    restriction_subs[restriction->m_name] = op_sym;
                 }
-                check_restriction(subs, restriction_subs, restriction, f_arg, x.base.base.loc);
+            } else {
+                LCOMPILERS_ASSERT(false);
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:683]: I would create a compiler error

File: integration_tests/template_add_02.f90 [hunk 4]
```diff
@@ -0,0 +1,59 @@
+module template_add_m
```
Comment (certik) [inline integration_tests/template_add_02.f90:1]: rename module names

File: src/lfortran/semantics/ast_common_visitor.h [hunk 5]
```diff
@@ -687,6 +687,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
     Vec<ASR::stmt_t*> *current_body = nullptr;
 
     // fields for generics
+    // TODO: simplify these fields
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:690]: remove

**Pull request merged**